### ISSUE TITLE
Inconsistent Results from NSVerify, ReluVal, Reluplex

### ISSUE THREAD

**Issue by tclayton0594** [sisl/NeuralVerification.jl#70]
Hi, I am opening this new issue as a follow-up discussion I had with @tomerarnon. It was brought to my attention that ReluVal may not be a complete solver, but to my knowledge NSVerify and Reluplex should be. For the input set, I defined upper and lower bound arrays for my 5-D input and then formed a hyperrectangle polytope. For the output set, I made 2 seperate: 1 for ReluVal and the other for NSVerify and Reluplex since they need a PC output. Since ranges form convex sets, I defined the PC with solely an upper bound of 287.1312. Here is a screenshot of how I setup the parameters for the solvers. 

![Reluplex Reluval Problem Setup](https://user-images.githubusercontent.com/32907889/62949292-ee84ff80-bdb3-11e9-8675-08eda873ddb4.PNG)

The network I am working with has an input size of 5, 3 hidden layers with 30 neurons each, and a single output neuron. Below is a screenshot of the results I got. 

![Reluplex Reluval Problem Results](https://user-images.githubusercontent.com/32907889/62949485-3f94f380-bdb4-11e9-98d1-78ce774bf0aa.PNG)

It is interesting that the counter example returned by Reluval is the midpoint of my input space, but it is unfortunately not a valid counter example. Both Reluplex and NSVerify use the same input/output types and are listed as complete, yet they gave me contradictory results. NSVerify indeed returned a valid counter example point. 

I also went a step further and narrowed my input domain around the violation found by NSVerify and ReluVal was able to find a valid counter example, but Reluplex still returned an incorrect holds. Any thoughts on this/is my formatting of the input and output spaces correct?

**Comment by tomerarnon** [comment]
Thanks for this @tclayton0594, I'm looking into it. Without your network I cannot reproduce this exactly. You can email me it if you'd like to (and indeed can) share it privately. At the moment I am just trying to determine what could have led to this failure path for each of the algorithms to help debug.

First I should note that what I said about reluval in #64 was wrong. Reluval is complete but "nonterminating". Reluval continues to split the input space as many times as `max_iter` will allow. If it cannot conclude either a :violated or a :holds at the end of that it should return :unknown, so what you're seeing may indeed be a bug. Although it is not printed in your reply, I understand that you tested the point and it definitely lies in Y? The reason it is the center of X is due to the marked line below that checks the midpoint of an interval. It seems like ReluVal is finding a :violated for you on the first iteration (therefore returning the midpoint of X).
```julia
function check_inclusion(reach::SymbolicInterval, output::AbstractPolytope, nnet::Network)
    reachable = symbol_to_concrete(reach)
    issubset(reachable, output) && return BasicResult(:holds)

    # Sample the middle point
    middle_point = (high(reach.interval) + low(reach.interval))./2
    y = compute_output(nnet, middle_point)
    ∈(y, output) || return CounterExampleResult(:violated, middle_point) # THIS LINE

    return BasicResult(:unknown)
end
```


My first guess of numerical instability in Reluplex/NSVerify is probably incorrect given your results. I'll keep looking there as well.

**Comment by tclayton0594** [comment]
Thanks, @tomerarnon for taking your time to verify. I am also verifying on my end as I was unaware that MATLAB did data manipulation by default, which may be the reason. Crazy your solvers were able to still find valid counter examples but within the next few hours I will have verified the results. Sorry for any confusion.

**Comment by tclayton0594** [comment]
I did verify the .nnet structure and still getting inconsistent results. I will continue to validate on my end since we would really like using some of your guys solver for verification of NNs for production if at all possible.

**Comment by tomerarnon** [comment]
Hi @tclayton0594, I wanted to follow up regarding whether you are able to share your network with us. Like I said, it's hard to follow the failure path without reproducing it. You can also reach me at tarnon@stanford.edu

**Comment by tomerarnon** [comment]
Hi @tclayton0594, no worries for the delay. I think I've just managed to recreate the Reluplex bug that you're seeing with my own network. I'll pursue the matter and get back to you. If you put together a sharable network still, I would really want to test with it as well.

As for numerical instability warnings from JuMP: unfortunately, there is no way to determine unequivocally whether the result is still valid. Subject to numerical instability, values used by the solver may be ±ϵ, so either a :holds or a :violated could be off by up to ϵ in those cases. In the case of a :violated, you could test the point and see if it is really invalid, and in the case of a dubious :holds, you could try tightening the constraint by some small number (making the problem slightly more conservative) to see if the result still holds.

**Comment by tclayton0594** [comment]
@tomerarnon , I got pulled onto another project that had a deadline of the 1st of the month but I am able to work on this again. I had another network I wanted to validate but I was running into the issue with ReluVal spitting out the midpoint as a counter example for many intervals, when in fact they were not counter examples. I am going to try and replicate the error on a dummy network. I had difficulty replicating the last time I tried but hopefully I will have more time to try this time. 

Thanks for all the help.
